Support Access-Control-Expose-Headers in CORS Handler

dims opened the pull request:
Our typical HTTP Response has a "Date" Header, if we don't add an
additional http header "Access-Control-Expose-Headers: Date" then
the browser based clients cannot use the Date HTTP Header.

Fixes #33231

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

``` release-note
When CORS Handler is enabled, we now add a new HTTP header named "Access-Control-Expose-Headers" with a value of "Date". This allows the "Date" HTTP header to be accessed from XHR/JavaScript.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33242)

<!-- Reviewable:end -->

nikhiljindal commented on pkg/apiserver/handlers.go:
```diff
@@ -402,8 +402,12 @@ func CORS(handler http.Handler, allowedOriginPatterns []*regexp.Regexp, allowedM
 					allowedHeaders = []string{"Content-Type", "Content-Length", "Accept-Encoding", "X-CSRF-Token", "Authorization", "X-Requested-With", "If-Modified-Since"}
 				}
+				if exposedHeaders == nil {
+					exposedHeaders = []string{"Date"}
```
Add a comment wherever we are adding the Date header saying if we remove that, we should remove this as well?

nikhiljindal commented:
Thanks, lg.
Reading the issue, it seems like this is waiting for @jwforres to try this out and confirm that this fixes the issue?

Also, needs rebase.

Let me know whenever this is ready to be merged and I can the LGTM label. Nothing blocking from me

nikhiljindal commented:
Also add a release note saying we are now adding `Access-Control-Expose-Headers` header?

jwforres commented:
yep sorry, I will try to get around to testing this tomorrow

On Tue, Sep 27, 2016 at 5:00 PM, Nikhil Jindal <notifications@github.com>
wrote:

> Also add a release note saying we are now adding
> Access-Control-Expose-Headers header?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33242#issuecomment-249996990>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7RsWEfcExePrprZ9xhFRrkBweTwHks5quYP7gaJpZM4KDdZx>
> .
>

jwforres commented:
@dims just verified with this patch in place that I can get the Date header from a cross-origin XHR object, thanks!

dims commented:
@jwforres thanks!

dims commented:
@nikhiljindal I added a release note, can you PTAL when you get a chance?

nikhiljindal commented:
Looks good. Thanks
Needs rebase

dims committed:
Support Access-Control-Expose-Headers in CORS Handler

Our typical HTTP Response has a "Date" Header, if we don't add an
additional http header "Access-Control-Expose-Headers: Date" then
the browser based clients cannot use the Date HTTP Header.

Fixes #33231

nikhiljindal commented:
Adding back LGTM
